keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,1071 @@
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
+    onRequiredIndexes: [0,1,2]
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a machine-learning researcher, I run an indexed job with an autoscale semantic
+set as a minimum number is 5 and a maximum number is 20.
+I want to mark the job as successful at least 5 indexes succeeded, even if 
+all indexes don't succeed.
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
+    onIndexesThreshold: 5
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 3
+
+As a machine-learning researcher from the [Story 2](#story-2), I want to mark the job
+as successful if 50% of all indexes succeed because I can not know how many indexes 
+will have in advance.
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
+    onIndexesThreshold: 50%
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Incorrect `.status.completedIndexes`
+
+If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessfulPolicy correctly.
+
+#### The Job object too big
+
+If we allow users to set all indexes (`0~10^5`) to `.spec.successfulPolicy.requiredIndexes`
+and don't limit the number of list sizes, there are cases that the size of a job object is too big.  
+
+In the worst case, allowed all indexes (`0~10^5`) are added to requiredIndexes.
+In that case, a `requiredIndexes` size is `SUM[9*10^n*(n+1)]+2+6≈5.6656MiB`, where:
+- `n` starts from `0` and goes up to `5`.
+- `1` of `n+1` means a separator that indicates `,`.
+- `2` is the sum of the index `0` and `,`.
+- `6` is the size of indexes `10^5`.
```

Comment:
@alculquicondor @mimowo 
This risk is similar to #3996. Do we need to change `requiredIndex` from `[]int` typed to `string` typed and then use the same format with [`.status.completedIndexes`](https://github.com/kubernetes/kubernetes/blob/93ec57734fbb97b33d6060af834a1dfe045fd914/pkg/apis/batch/types.go#L390-L397)?

Or, it might be better to keep using `[]int` typed and limit the list sizes since `requiredIndexes` is modified by end-users.

wdyt?

Reply:
A general driver-worker setup is the latter. I'm fine with having `min` and `max`.

@vsoch Do you have the former scenarios (index 1, 3, 7 to finish)?

Reply:
If any others don't have the former scenarios, I will move the current design and risks to the alternative section.

Reply:
I do not, but theoretically we could in the future with the Flux Operator if we allow for a more complex topology that launches more than one parent at the top level (and thus successful completion would rely on that set of parents). So I'm +1 to have the success case allow a set, where the set is allowed to be one or more. @alculquicondor's suggestion makes sense, but it might be a more confusing interaction for this case of having just one, and (in the future) would be limiting in allowing for more unique descriptions. Perhaps you could instead do what we do for hostnames, and allow for a set, and allow the set to include ranges? E.g.,:

```yaml
# The Flux Operator current design - we are done when index 0 is done.
requiredIndexesInterval: [0]

# A more complex topology that has several parents, in a range, flux-sample-0, flux-sample-1, and flux-sample-2
requiredIndexesInterval: [0-2]

# A more complex topology that has several parents, but not in a range flux-sample-0, flux-sample-5, flux-sample-10
requiredIndexesInterval: [0,5,10]

# The same, but with a combination of one off and ranges
requiredIndexesInterval: [0-2,5,10]
```
For the third, I think this could be something that happens if there is some third variable like topology that we want to map lead brokers to. E.g., imagine that you have 100 pods, each assigned to one node for simplicity, and the topology chosen (if possible) says "put indices that are physically closer together also numerically close." In this case you'd maybe not want parents to be 0-3, but to spread out the leaders to reflect that design.

And finally, looking at the above you might need to allow string OR int and then parse appropriately. At least for config files for Flux and Slurm that syntax (without additional string declaration) would be allowed.

Reply:
Currently, we use the format in `.status.completedIndexes`:
https://github.com/kubernetes/kubernetes/blob/93ec57734fbb97b33d6060af834a1dfe045fd914/pkg/apis/batch/types.go#L390-L397

> For the third, I think this could be something that happens if there is some third variable like topology that we want to map lead brokers to. 

@vsoch Why do you use discrete values? Can we use `[0-3]`, and then put each leader at different topologies?

slurm topology example: https://slurm.schedmd.com/topology.html

Reply:
Thanks for the great clarifications.

> if it's easier to manually control topology, the range should be fine here (but I am skeptical of this claim)

I believe that we can use Inter-pod anti-affinity based on the `batch.kubernetes.io/job-completion-index` label.

https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity

https://github.com/kubernetes/kubernetes/issues/117986

However, the label will be added to pods since v1.28. So the label isn't added to pods in v1.27.

```yaml
...
spec:
  completions: 10
  parallelism: 10
  completionMode: Indexed
  template:
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: batch.kubernetes.io/job-completion-index
                operator: In
                values: ["0", "1", "2"] // leaders
            topologyKey: topology.kubernetes.io/zone
...
```

Reply:
> But since we already have precedent with completedIndexes, using the compressed format should be good. However, I would add validation to limit the size of the field to up to 64Ki or 100Ki (more than plenty, IMHO)

I'm ok with the compressed format and the validation.

Reply:
Anyway, I will update the API design.

Reply:
Done.